File: backend/api/routers/exercises.py
"""
Exercises API Router
运动/功法 API 路由 - MVP
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import Dict, Any, Optional, List

from api.database import get_db
from api.services.exercise_service import ExerciseService
from pydantic import BaseModel


router = APIRouter(prefix="/exercises", tags=["exercises"])


# Response Models
class ExerciseListResponse(BaseModel):
    code: int
    message: str
    data: Dict[str, Any]


class ExerciseDetailResponse(BaseModel):
    code: int
    message: str
    data: Dict[str, Any]


@router.get("", response_model=ExerciseListResponse)
async def get_exercises(
    skip: int = Query(0, ge=0, description="跳过数量"),
    limit: int = Query(20, ge=1, le=100, description="限制数量"),
    exercise_type: Optional[str] = Query(None, description="运动类型筛选"),
    difficulty_level: Optional[str] = Query(None, description="难度级别筛选"),
    search: Optional[str] = Query(None, description="搜索关键词"),
    db: Session = Depends(get_db)
) -> Dict[str, Any]:
    """
    获取运动列表

    支持按运动类型、难度级别筛选和关键词搜索
    """
    service = ExerciseService()
    exercises, total = service.get_exercises_list(
        db=db,
        skip=skip,
        limit=limit,
        exercise_type=exercise_type,
        difficulty_level=difficulty_level,
        search=search
    )

    return {
        "code": 0,
        "message": "success",
        "data": {
            "total": total,
            "items": [
                {
                    "id": ex.id,
                    "name": ex.name,
                    "name_en": ex.name_en,
                    "description": ex.description,
                    "exercise_type": ex.exercise_type,
                    "difficulty_level": ex.difficulty_level,
                    "duration_seconds": ex.duration_seconds,
                    "video_url": ex.video_url,
                    "image_url": ex.image_url,
                    "target_constitutions": ex.target_constitutions,
                    "view_count": ex.view_count
                }
                for ex in exercises
            ]
        }
    }


@router.get("/types", response_model=Dict[str, Any])
async def get_exercise_types() -> Dict[str, Any]:
    """
    获取运动类型列表
    """
    service = ExerciseService()
    types = service.get_exercise_types(None)

    return {
        "code": 0,
        "message": "success",
        "data": {
            "types": types
        }
    }


@router.get("/constitution/{constitution_code}", response_model=ExerciseListResponse)
async def get_exercises_by_constitution(
    constitution_code: str,
    limit: int = Query(20, ge=1, le=100, description="限制数量"),
    db: Session = Depends(get_db)
) -> Dict[str, Any]:
    """
    根据体质获取推荐运动
    """
    service = ExerciseService()
    exercises = service.get_exercises_by_constitution(
        constitution_code=constitution_code,
        db=db,
        limit=limit
    )

    return {
        "code": 0,
        "message": "success",
        "data": {
            "constitution": constitution_code,
            "total": len(exercises),
            "items": [
                {
                    "id": ex.id,
                    "name": ex.name,
                    "name_en": ex.name_en,
                    "description": ex.description,
                    "exercise_type": ex.exercise_type,
                    "difficulty_level": ex.difficulty_level,
                    "duration_seconds": ex.duration_seconds,
                    "video_url": ex.video_url,
                    "image_url": ex.image_url,
                    "target_constitutions": ex.target_constitutions,
                    "view_count": ex.view_count
                }
                for ex in exercises
            ]
        }
    }


@router.get("/routine/{constitution_code}", response_model=Dict[str, Any])
async def get_personalized_routine(
    constitution_code: str,
    db: Session = Depends(get_db)
) -> Dict[str, Any]:
    """
    获取个性化运动方案（早中晚）
    """
    service = ExerciseService()
    routine = service.get_personalized_routine(
        constitution_code=constitution_code,
        db=db
    )

    return {
        "code": 0,
        "message": "success",
        "data": {
            "constitution": constitution_code,
            "routine": {
                "morning": [
                    {
                        "id": ex.id,
                        "name": ex.name,
                        "exercise_type": ex.exercise_type,
                        "difficulty_level": ex.difficulty_level,
                        "duration_seconds": ex.duration_seconds,
                        "video_url": ex.video_url,
                        "image_url": ex.image_url
                    }
                    for ex in routine.get("morning", [])
                ],
                "afternoon": [
                    {
                        "id": ex.id,
                        "name": ex.name,
                        "exercise_type": ex.exercise_type,
                        "difficulty_level": ex.difficulty_level,
                        "duration_seconds": ex.duration_seconds,
                        "video_url": ex.video_url,
                        "image_url": ex.image_url
                    }
                    for ex in routine.get("afternoon", [])
                ],
                "evening": [
                    {
                        "id": ex.id,
                        "name": ex.name,
                        "exercise_type": ex.exercise_type,
                        "difficulty_level": ex.difficulty_level,
                        "duration_seconds": ex.duration_seconds,
                        "video_url": ex.video_url,
                        "image_url": ex.image_url
                    }
                    for ex in routine.get("evening", [])
                ]
            }
        }
    }


@router.get("/plan/{constitution_code}/{week}", response_model=Dict[str, Any])
async def get_exercise_plan_week(
    constitution_code: str,
    week: int,
    db: Session = Depends(get_db)
) -> Dict[str, Any]:
    """
    获取周运动计划
    """
    service = ExerciseService()
    plan = service.get_exercise_plan_week(
        constitution_code=constitution_code,
        week=week,
        db=db
    )

    return {
        "code": 0,
        "message": "success",
        "data": plan
    }


@router.get("/{exercise_id}", response_model=ExerciseDetailResponse)
async def get_exercise_detail(
    exercise_id: str,
    db: Session = Depends(get_db)
) -> Dict[str, Any]:
    """
    获取运动详情
    """
    service = ExerciseService()
    exercise = service.get_exercise_by_id(exercise_id, db)

    if not exercise:
        raise HTTPException(status_code=404, detail="Exercise not found")

    # 增加浏览次数
    service.increment_view_count(exercise_id, db)

    return {
        "code": 0,
        "message": "success",
        "data": {
            "id": exercise.id,
            "name": exercise.name,
            "name_en": exercise.name_en,
            "description": exercise.description,
            "exercise_type": exercise.exercise_type,
            "difficulty_level": exercise.difficulty_level,
            "target_constitutions": exercise.target_constitutions,
            "video_url": exercise.video_url,
            "image_url": exercise.image_url,
            "step_images": exercise.step_images,
            "instructions": exercise.instructions,
            "duration_seconds": exercise.duration_seconds,
            "repetitions": exercise.repetitions,
            "target_body_areas": exercise.target_body_areas,
            "contraindications": exercise.contraindications,
            "caution_notes": exercise.caution_notes,
            "benefits": exercise.benefits,
            "tags": exercise.tags,
            "view_count": exercise.view_count
        }
    }


@router.post("/{exercise_id}/view", response_model=Dict[str, Any])
async def increment_exercise_view(
    exercise_id: str,
    db: Session = Depends(get_db)
) -> Dict[str, Any]:
    """
    增加运动浏览次数
    """
    service = ExerciseService()
    success = service.increment_view_count(exercise_id, db)

    if not success:
        raise HTTPException(status_code=404, detail="Exercise not found")

    return {
        "code": 0,
        "message": "View count incremented",
        "data": {"exercise_id": exercise_id}
    }

File: frontend/src/api/exercises.js
/**
 * 运动/功法相关 API - MVP
 */
import { get, post, put } from '@/utils/request.js'

/**
 * 获取运动列表
 * @param {Object} params - 查询参数
 * @param {number} params.skip - 跳过数量
 * @param {number} params.limit - 限制数量
 * @param {string} params.exercise_type - 运动类型筛选
 * @param {string} params.difficulty_level - 难度级别筛选
 * @param {string} params.search - 搜索关键词
 * @returns {Promise<Object>} 返回 {total, items}
 */
export function getExercises(params = {}) {
  return get('/api/v1/exercises', params).then(res => {
    if (res.code === 0) return res.data
    throw new Error(res.message || '获取运动列表失败')
  })
}

/**
 * 获取运动类型列表
 * @returns {Promise<Array>} 返回类型列表
 */
export function getExerciseTypes() {
  return get('/api/v1/exercises/types').then(res => {
    if (res.code === 0) return res.data.types
    throw new Error(res.message || '获取运动类型失败')
  })
}

/**
 * 根据体质获取推荐运动
 * @param {string} constitutionCode - 体质代码
 * @param {Object} params - 查询参数
 * @param {number} params.limit - 限制数量
 * @returns {Promise<Object>} 返回 {constitution, total, items}
 */
export function getExercisesByConstitution(constitutionCode, params = {}) {
  return get(`/api/v1/exercises/constitution/${constitutionCode}`, params).then(res => {
    if (res.code === 0) return res.data
    throw new Error(res.message || '获取推荐运动失败')
  })
}

/**
 * 获取个性化运动方案（早中晚）
 * @param {string} constitutionCode - 体质代码
 * @returns {Promise<Object>} 返回 {constitution, routine: {morning, afternoon, evening}}
 */
export function getPersonalizedRoutine(constitutionCode) {
  return get(`/api/v1/exercises/routine/${constitutionCode}`).then(res => {
    if (res.code === 0) return res.data
    throw new Error(res.message || '获取运动方案失败')
  })
}

/**
 * 获取周运动计划
 * @param {string} constitutionCode - 体质代码
 * @param {number} week - 第几周
 * @returns {Promise<Object>} 返回 {week, constitution, daily_plan}
 */
export function getExercisePlanWeek(constitutionCode, week) {
  return get(`/api/v1/exercises/plan/${constitutionCode}/${week}`).then(res => {
    if (res.code === 0) return res.data
    throw new Error(res.message || '获取周计划失败')
  })
}

/**
 * 获取运动详情
 * @param {string} exerciseId - 运动ID
 * @returns {Promise<Object>} 返回运动详情
 */
export function getExerciseDetail(exerciseId) {
  return get(`/api/v1/exercises/${exerciseId}`).then(res => {
    if (res.code === 0) return res.data
    throw new Error(res.message || '获取运动详情失败')
  })
}

/**
 * 增加运动浏览次数
 * @param {string} exerciseId - 运动ID
 * @returns {Promise<Object>}
 */
export function incrementViewCount(exerciseId) {
  return post(`/api/v1/exercises/${exerciseId}/view`).then(res => {
    if (res.code === 0) return res.data
    throw new Error(res.message || '操作失败')
  })
}

export default {
  getExercises,
  getExerciseTypes,
  getExercisesByConstitution,
  getPersonalizedRoutine,
  getExercisePlanWeek,
  getExerciseDetail,
  incrementViewCount
}
